ddtrace/opentelemetry/span_test.go

```diff
@@ -6,79 +6,305 @@
 package opentelemetry
 
 import (
+	"bytes"
 	"context"
 	"fmt"
 	"io"
 	"net/http"
 	"testing"
 	"time"
 
+	"go.opentelemetry.io/otel/attribute"
+
 	"github.com/stretchr/testify/assert"
+	"github.com/tinylib/msgp/msgp"
 	"go.opentelemetry.io/otel"
-	"go.opentelemetry.io/otel/attribute"
 	"go.opentelemetry.io/otel/codes"
 
 	"gopkg.in/DataDog/dd-trace-go.v1/ddtrace/tracer"
 	"gopkg.in/DataDog/dd-trace-go.v1/internal/httpmem"
 )
 
-func TestSpanMethods(t *testing.T) {
-	testData := struct {
-		env, srv, oldOp, newOp string
-		tags                   [][]string // key - tags[0][0], value - tags[0][1]
-	}{
-		env:   "test_env",
-		srv:   "test_srv",
-		oldOp: "old_op",
-		newOp: "new_op",
-		tags:  [][]string{{"tag_1", "tag_1_val"}, {"opt_tag_1", "opt_tag_1_val"}}}
-	done := make(chan struct{})
-	var payload string
+func mockTracerProvider(t *testing.T, opts ...tracer.StartOption) (tp *TracerProvider, payloads chan string, cleanup func()) {
+	payloads = make(chan string)
 	s, c := httpmem.ServerAndClient(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		switch r.URL.Path {
 		case "/v0.4/traces":
+			if h := r.Header.Get("X-Datadog-Trace-Count"); h == "0" {
+				return
+			}
 			buf, err := io.ReadAll(r.Body)
-			if err != nil {
+			if err != nil || len(buf) == 0 {
+				t.Log("Test agent: Error receiving traces")
 				t.Fail()
 			}
-			payload = fmt.Sprintf("%s", buf)
-			done <- struct{}{}
+			var js bytes.Buffer
+			msgp.UnmarshalAsJSON(&js, buf)
+			select {
+			case payloads <- js.String():
+			default:
+				t.Log("Test agent: no one to receive payloads")
+			}
 		}
 		w.WriteHeader(200)
 	}))
-	defer s.Close()
+	opts = append(opts, tracer.WithHTTPClient(c))
+	tp = NewTracerProvider(opts...)
+	otel.SetTracerProvider(tp)
+	return tp, payloads, func() {
+		tp.Shutdown()
+		s.Close()
+	}
+}
+
+func waitForPayload(ctx context.Context, t *testing.T, payloads chan string) string {
+	var p string
+	select {
+	case <-ctx.Done():
+		t.Fatal("timed out waiting for traces")
+	case p = <-payloads:
+		break
+	}
+	return p
+}
+
+func TestSpanSetName(t *testing.T) {
+	assert := assert.New(t)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
 
-	otel.SetTracerProvider(NewTracerProvider(
-		tracer.WithEnv(testData.env),
-		tracer.WithHTTPClient(c),
-		tracer.WithService(testData.srv)))
+	_, payloads, cleanup := mockTracerProvider(t)
 	tr := otel.Tracer("")
+	defer cleanup()
+
+	_, sp := tr.Start(ctx, "OldName")
+	sp.SetName("NewName")
+	sp.End()
+
+	tracer.Flush()
+	p := waitForPayload(ctx, t, payloads)
+	assert.Contains(p, "NewName")
+}
+
+func TestSpanEnd(t *testing.T) {
```

Comment:
I remember this being refactored in https://github.com/DataDog/dd-trace-go/pull/1701/files#diff-f8e8831aaae7ae21f5494aee61a705d540c32154c5ebd4df88f42a3b736ebae6 based on some of my comments. Maybe needs a rebase? I imagine this delta will just disappear at that point, unless I'm mixing something up.